tfexec/terraform_state_replace_provider_test.go

```diff
@@ -0,0 +1,290 @@
+package tfexec
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/go-version"
+)
+
+func TestTerraformCLIStateReplaceProvider(t *testing.T) {
+	state := NewState([]byte("dummy state"))
+	updatedState := NewState([]byte("updated dummy state"))
+
+	// mock writing state to a temporary file.
+	runFunc := func(args ...string) error {
+		for _, arg := range args {
+			if strings.HasPrefix(arg, "-state=") {
+				stateFile := arg[len("-state="):]
+				err := os.WriteFile(stateFile, updatedState.Bytes(), 0600)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}
+
+	versionRunFunc := func(args ...string) error {
+		return nil
+	}
+
+	cases := []struct {
+		desc         string
+		mockCommands []*mockCommand
+		state        *State
+		source       string
+		destination  string
+		opts         []string
+		updatedState *State
+		ok           bool
+	}{
+		{
+			desc: "no opts",
+			mockCommands: []*mockCommand{
+				{
+					args:     []string{"terraform", "version"},
+					argsRe:   regexp.MustCompile(`^terraform version$`),
+					runFunc:  versionRunFunc,
+					stdout:   "Terraform v1.5.0\n",
+					exitCode: 0,
+				},
+				{
+					args:     []string{"terraform", "state", "replace-provider", "registry.terraform.io/-/null", "registry.terraform.io/hashicorp/null"},
+					argsRe:   regexp.MustCompile(`^terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null$`),
+					runFunc:  runFunc,
+					exitCode: 0,
+				},
+			},
+			state:        nil,
+			source:       "registry.terraform.io/-/null",
+			destination:  "registry.terraform.io/hashicorp/null",
+			updatedState: nil,
+			ok:           true,
+		},
+		{
+			desc: "failed to run terraform state replace-provider",
+			mockCommands: []*mockCommand{
+				{
+					args:     []string{"terraform", "version"},
+					argsRe:   regexp.MustCompile(`^terraform version$`),
+					runFunc:  versionRunFunc,
+					stdout:   "Terraform v1.5.0\n",
+					exitCode: 0,
+				},
+				{
+					args:     []string{"terraform", "state", "replace-provider", "registry.terraform.io/-/null", "registry.terraform.io/hashicorp/null"},
+					argsRe:   regexp.MustCompile(`^terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null$`),
+					runFunc:  runFunc,
+					exitCode: 1,
+				},
+			},
+			state:        nil,
+			source:       "registry.terraform.io/-/null",
+			destination:  "registry.terraform.io/hashicorp/null",
+			updatedState: nil,
+			ok:           false,
+		},
+		{
+			desc: "with unsupported Terraform version",
+			mockCommands: []*mockCommand{
+				{
+					args:     []string{"terraform", "version"},
+					argsRe:   regexp.MustCompile(`^terraform version$`),
+					runFunc:  versionRunFunc,
+					stdout:   "Terraform v0.12.99\n",
+					exitCode: 1,
```

Comment:
When the exit code of the terraform version command is 1,  it doesn't work as expected because the function returns before checking the version constraint.